### PR TITLE
Add required user-agent to crates-io lib api (breaking change)

### PR DESCRIPTION
**Pull request by FGRibreau**
Since cargo.io now requires a User-Agent header (see message below), I've updated the crates-io crate accordingly. Hope this will be approved and generate a new release of crates-io :)

This is a breaking change because there is an contract change on (added `user_agent: Option<String>`):
- `Registry::new(host: String, token: Option<String>, user_agent: Option<String>)`
- `Registry::new_handle(host: String, token: Option<String>, user_agent: Option<String>, handle: Easy)`

Message from crates.io:

```
We require that all requests include a `User-Agent` header. 
To allow us to determine the impact your bot has on our service, we ask that your user agent actually identify your bot, and not just report the HTTP client library you're using.  
Including contact information will also reduce the chance that we will need to take action against your bot.

Bad:
User-Agent: reqwest/0.9.1

Better:
User-Agent: my_crawler

Best:
User-Agent: my_crawler (my_crawler.com/info)
User-Agent: my_crawler (help@my_crawler.com)

If you believe you've received this message in error, please email help@crates.io and include the request id 7da610ae-264a-4738-a979-c53465fa8c99.
```

### CODE REVIEW

File: src/crates-io/lib.rs [hunk 1]
```diff
@@ -127,14 +130,15 @@ struct Crates {
...
+    pub fn new(host: String, token: Option<String>, user_agent: Option<String>) -> Registry {
+        Registry::new_handle(host, token, user_agent, Easy::new())
     }
 
-    pub fn new_handle(host: String, token: Option<String>, handle: Easy) -> Registry {
+    pub fn new_handle(host: String, token: Option<String>, user_agent: Option<String>, handle: Easy) -> Registry {
         Registry {
             host,
             token,
+            user_agent: userAgent,
```
Comment (ehuss) [inline src/crates-io/lib.rs]: Looks like a typo?

Comment (FGRibreau) [reply src/crates-io/lib.rs]: fixed thanks :)

**Comment by FGRibreau** [comment]
> Also, I'm thinking it might be better to not use an Option, and require an explicit value.

@ehuss you're right indeed, I removed the Option :)

**Comment by ehuss** [comment]
Can you also update `src/cargo/ops/registry.rs` as mentioned above?

**Comment by ehuss** [comment]
Hm, I think I wasn't clear.  There is custom handling for the user agent in `configure_http_handle`.  That will need to be handled, and somehow shared. Unfortunately this stuff is used all over, so it may not be so simple to update.  Perhaps it would be best just to update `Registry::new` and leave `Registry::new_handle` unchanged?  Cargo doesn't use `Registry::new`, and maybe a doc comment could be added to `new_handle` informing the user they must set the user-agent header.

**Comment by alexcrichton** [comment]
This has gone stale for some time now so I'm going to close this, but feel free to resubmit of course!

**Pull request closed without merging**